prototypes/notebook-generator/code_template.py

```diff
@@ -0,0 +1,93 @@
+from string import Template
+from libertem.analysis.base import Analysis
+
+
+class CodeTemplate():
+    def __init__(self, connection, dataset, compound_analysis):
+        self.conn = connection
+        self.ds = dataset
+        self.compound_analysis = compound_analysis
+
+    def format_template(self, template, data):
+        template = "\n".join(template)
+        return Template(template).substitute(data)
+
+    def dataset(self):
+        ds_type = self.ds['type']
+        ds_params = self.ds['params']
+        temp_ds = ['params = $params',
+                   'ds = ctx.load("$type", **params)']
+        data = {'type': ds_type, 'params': ds_params}
+        return self.format_template(temp_ds, data)
+
+    def dependency(self):
+        temp_dep = ["import matplotlib.pyplot as plt",
+                    "import libertem.api as lt",
+                    "import numpy as np",
+                    "from libertem.analysis.getroi import get_roi",
+                    "import numpy as np"]
+
+        return '\n'.join(temp_dep)
+
+    def initial_setup(self):
+        return "%matplotlib nbagg"
+
+    def connection(self):
+        docs = ["# Connection"]
+        if (self.conn['type'] == "cluster"):
+            docs.append("***more about cluster conn")
+            temp_conn = ['cluster = executor.dask.DaskJobExecutor.connect("$conn_url")',
+                         "ctx = lt.Context(executor=cluster)"]
+            data = {'conn_url': self.conn['url']}
+            ctx = self.format_template(temp_conn, data)
+            docs = '\n'.join(docs)
+            return ctx, docs
+        else:
+            docs.append("**more about local conn**")
+            ctx = "ctx = lt.Context()"
+            docs = '\n'.join(docs)
+            return ctx, docs
+
+    def analysis(self):
+
+        temp_docs = ["# $analysis",
+                     "**description**"]
+
+        temp_roi = ["roi_params = $roi_params",
+                    "roi = get_roi(roi_params, ds.shape.nav)"]
+
+        temp_analysis = ["${short}_analysis = ctx.$analysis_api($params)",
+                         "$roi",
+                         "udf = ${short}_analysis.get_udf()",
+                         "${short}_result = ctx.run_udf(ds, udf, roi, progress=True)"]
+
+        form_analysis = []
+
+        for analysis in self.compound_analysis:
+
+            an_type = analysis['analysisType']
+            an_params = analysis['parameters']
+            cls = Analysis.get_analysis_by_type(an_type)
+            helperCls = cls.get_template_helper()
+            helper = helperCls()
+
+            analysis_api = helper.api
+            short = helper.short_name
+
+            plot_ = helper.get_plot()
+            docs_ = self.format_template(temp_docs, {'analysis': an_type})
+
+            if 'roi' in an_params.keys():
+                data = {'roi_params': an_params}
```

Comment:
It's the parameters for `get_roi()`, example [notebook](https://github.com/LiberTEM/LiberTEM/blob/01f3569f9166c60550f1a20ddb32a2e2cb890dfe/prototypes/notebook-generator/comp_analysis_notebook.ipynb).

Reply:
Ok, but if the analysis has more parameters than just the `roi`, how is that handled? I think this may be the case for one of the FFT analyses? `SumfftAnalysis`?

Reply:
does the `sumfftAnalysis` have `roi` as parameter?

Reply:
Yes, as it is a subclass of the `SumAnalysis`. IIRC all analyses that have a `sig`-shaped result support ROIs, and you can check by using the ROI dropdown in the GUI

Reply:
![fft_analysis](https://user-images.githubusercontent.com/59721339/84811400-6e11d980-b02a-11ea-8063-9fa20d714e37.png)
ROI dropdown is missing

Reply:
Ah, so it was never added to the FFT analysis in the GUI. But you can use your imagination: 

![image](https://user-images.githubusercontent.com/5778/84815933-0a7cb180-b014-11ea-8586-f00a0088af99.png)

These are the analysis parameters; if there was a `roi`, there would also be a `roi` key under parameters, in addition to the `real_*` ones here:

![image](https://user-images.githubusercontent.com/5778/84816144-5596c480-b014-11ea-9e4b-45d98fb12b82.png)